### ISSUE TITLE
@AfterViews in DialogFragment

### ISSUE THREAD

**Issue by dhleong** [excilys/androidannotations#486]
I have a DialogFragment that I'm planning to use in both Dialog mode and embedded mode. I have an `AfterViews` method that doesn't seem to get called when the fragment is used as a Dialog. This also means, of course, that @Click methods, etc. don't fire.

I'm using the 3.0 snapshot.

My fragment looks something like the following, because I want to use dialog buttons where available, and to avoid crashes that may be support library-related:

```
// members, etc.

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
            Bundle savedInstanceState) {
        
        // in dialog mode... don't do anything (I guess) to prevent 
        //  "requestFeature() must be called before adding content" exception
        if (getDialog() != null)
            return null;
        
        View view = inflater.inflate(R.layout.frag_create_post, null);
        
        prepareView(view); // because @AfterViews isn't working
        
        return view;
    }
    
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {

        final View view = LayoutInflater.from(getActivity())
                .inflate(R.layout.frag_create_post, null);
        prepareView(view); // because @AfterViews isn't working
        
        AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());
        builder
            .setView(view)
            .setPositiveButton(R.string.share, this)
            .setNegativeButton(R.string.cancel, null);
        
        return builder.create();
    }
```

**Comment by taar1** [comment]
Are you guys going to work on a solution for the DialogFragment? We would be needing this for our application :-)

**Comment by dodgex** [comment]
hmm interesting. i have a DialogFragment where everything works just fine. i also use it as dialog (on tablet) or as fragment inside an activity (for phones). (tested on api 14, 19 and 22 devices/avd)

**Comment by WonderCsabo** [comment]
@dodgex @taar1 can you show examples how do you use `@AfterViews`?

**Comment by dodgex** [comment]
i just have 
```java
import android.app.DialogFragment;
// more imports
@EFragment(resName = "fragment_alarmeditor")
public class AlarmEditorFragment extends DialogFragment  {
// other code using AA annotations
}
```
to open the fragment i have
```java
if (PlayerApp.isDualPaneLayout()) {
	// opens the dialog
	AlarmEditorFragment_.builder().alarmInfo(info).build().callback(AlarmclockFragment.this).show(getFragmentManager(), "alarmeditor_dialog");
} else {
	// the activity contains the fragment in its layout
	AlarmEditorActivity_.intent(AlarmclockFragment.this).alarmInfo(info).startForResult(EDIT_ALARM);
}
```

**Comment by WonderCsabo** [comment]
@dodgex and do you use `@AfterViews`?

**Comment by dodgex** [comment]
`@AfterViews` and `@AfterInject`

**Comment by WonderCsabo** [comment]
Well, i am closing this done. Please reopen if this is really a defect.

**Comment by OskJonsdottir** [comment]
As mentioned above by @pyricau, the javadoc of `FragmentDialog` states "Implementations should override this class and implement onCreateView(LayoutInflater, ViewGroup, Bundle) to supply the content of the dialog."

But the sentence right after that one reads "**Alternatively, they can override onCreateDialog(Bundle) to create an entirely custom dialog, such as an AlertDialog, with its own content**".

When @dodgex says "I have a `DialogFragment` where everything works just fine", she or he should specify that it is "a `DialogFragment` that does NOT implement `onCreateDialog`". If understand the original sense of this issue, the problem reported here is that there is no way to use Android Annotations if you want to create a `DialogFragment` that implements `onCreateDialog()`. 

If you annotate your `FragmentDialog` with `@EFragment`, then `onCreateView()` is automatically implemented in the annotated class to return the content view described by the layout value passed to the `@EFragment` annotation. If you class also implements `onCreateDialog()`, an exception is thrown during creation.

**Comment by WonderCsabo** [comment]
Thanks for the detailed description. Sorry, but i am little bit outside of the context of the issue.

@OskJonsdottir what is you are suggesting to solve this?

**Comment by OskJonsdottir** [comment]
I have the feeling that this would be quite complicated to fix / implement. One would like to have access to the layout view and views referenced by id in the usual way. Something like this:
```java
    @EFragment(R.layout.password_dialog)
    public class PasswordDialogFragment extends DialogFragment {
        private PasswordEnteredListener listener;
    
        @ViewById(R.id.password_dialog_pwdEditText)
        EditText editText;
    
        @Override
        public Dialog onCreateDialog(Bundle savedInstanceState) {
            AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());
            builder.setView(getContentView())   //THIS getContentView() METHOD DOES NOT EXIST...
                    .setMessage(getActivity().getString(R.string.some_title)
                            .setNegativeButton("Cancel", null)
                            .setPositiveButton("OK", new DialogInterface.OnClickListener() {
                                @Override
                                public void onClick(DialogInterface dialog, int which) {
                                    if (listener != null) {
                                        listener.passwordEntered(editText.getText().toString());
                                    }
                                }
                            });
            return builder.create();
        }
    
        public void setListener(PasswordEnteredListener listener) {
            this.listener = listener;
        }
    }
```

But there is no way to access the contentView_ inflated by the annotated class, and worst of all, the method that inflates it in the current implementation of `@EFragment` is `onCreateView()`, which is called _after_ `onCreateDialog()`.

**Comment by WonderCsabo** [comment]
I think we should create an `@EDialogFragment` annotation, which will support the lifecycle of dialogs instead of normal `Fragment`s. Hacking support into `@EFragment` indeed shold require some hacks.

**Comment by OskJonsdottir** [comment]
So what would be your suggested syntax?

**Comment by WonderCsabo** [comment]
What you posted https://github.com/excilys/androidannotations/issues/486#issuecomment-110398332, just replace `@EFragment` with `@EDialogFragment`. What do you think?

**Comment by OskJonsdottir** [comment]
Well, that would be convenient to use, at least, I don't know how difficult it would be to implement. I wonder if the `getContentView()` method matches the general style of the Android Annotations framework. Perhaps another annotation might be created for that purpose, so that one could write:
```java
@ContentView
View contentView;
```
and then later
```java
        AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());
        builder.setView(contentView)
```
etc.

**Comment by WonderCsabo** [comment]
@OskJonsdottir maybe i am still missing some details. What should the generated class should look like?

**Comment by OskJonsdottir** [comment]
@WonderCsabo. I only just started using Android Annotations. I don't really know what the class should look like. :-)

**Comment by WonderCsabo** [comment]
@OskJonsdottir let me rephrase my question: in a `DialogFragment`, in what method we set the layout? What method should be used to inject the views?

**Comment by OskJonsdottir** [comment]
@WonderCsabo. If I'm not wrong, when you create a `DialogFragment` that implements `onDialogCreate()` what you are expected to create is full-blown `Dialog`. So you typically pass the layout to the `AlertDialog` builder that builds the `Dialog` that you return
```java
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());
        builder.setView(contentView)
        ...
        return builder.create();
    }
```
But if you implement `onCreateView()`, you return the `View` directly. 
```java
public View OnCreateView (LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
    View v = inflater.Inflate(R.Layout.some_view, container, false);
    return v;
}
```
Does that answer your question?

**Comment by OskJonsdottir** [comment]
Keep in mind, however, that you can, in principle, implement _both_ methods, as long as your implementation of `onCreateDialog()` invokes `super.onCreateDialog()` and eventually returns the `Dialog` obtained that way, i.e.,
```java
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        Dialog d = super.onCreateDialog(savedInstanceState);
        //do something to the Dialog
        //...
        return d;
    }

    public View OnCreateView (LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View v = inflater.Inflate(R.Layout.some_view, container, false);
        return v;
    }
```

**Comment by OskJonsdottir** [comment]
http://stackoverflow.com/a/19302991

**Comment by WonderCsabo** [comment]
Thanks, it is more clear now. I think we cannot help this if you want to create the `Dialog` instance yourself, because there is no time point we can create our layout.

The only thing we can do is injecting the views after `onCreateDialog` is called...

**Comment by OskJonsdottir** [comment]
Thanks @WonderCsabo, that is the impression I got. I guess the workaround is just to create the whole layout in xml and not use `AlertDialog.Builder`.

**Comment by WonderCsabo** [comment]
When you are implement https://github.com/excilys/androidannotations/issues/486#issuecomment-110672535, which method is called first, `onCreateDialog` or `onCreateView`?

**Comment by OskJonsdottir** [comment]
`onCreateDialog()` is called first, then `onCreateView()`.

**Comment by WonderCsabo** [comment]
OK. I think we cannot help this, then. :cry: You can try to leverage the view setup to an `@AfterViews` method, and do not create the `Dialog` instance yourself, as explained in https://github.com/excilys/androidannotations/issues/486#issuecomment-110672535. This way you can use `@EFragment`.

**Comment by OskJonsdottir** [comment]
@WonderCsabo thanks anyway. At least this thread can serve as documentation for other people that might want to try this in the future, to warn them that it is not possible to combine `onCreateDialog()` with Android Annotations.

**Comment by WonderCsabo** [comment]
I added this info the the [wiki](https://github.com/excilys/androidannotations/wiki/Enhance-Fragments#note-on-dialogfragments).

**Comment by OlegDokuka** [comment]
Better way to solve the problem: 
```java 
@EFragment
public class BoilerStateFragment extends DialogFragment {
    private View view;

    @ViewById(R.id.pager)
    ViewPager viewPager;
    @ViewById
    TabLayout tabLayout;

    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        final AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());
        final LayoutInflater inflater = getActivity().getLayoutInflater();
        view = inflater.inflate(R.layout.fragment_boiler_state, null);

        builder.setView(view)
                .setPositiveButton("OK", (dialog, id) -> {
                    // sign in the user ...
                })
                .setNegativeButton("CANCEL", (dialog, id) -> BoilerStateFragment.this.getDialog().cancel());

        return builder.create();
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return view;
    }

    @AfterViews
    void calledAfterViewInjection() {
       //some code
    }
}
```

**Comment by njuejohn** [comment]
After 4 days of hard work, this is the solution.

`@EFragment(R.layout.activity_location_view)
public class LocationViewActivity extends DialogFragment{ 

private View view;
//Other variables

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return view;
    }

 @AfterViews
    void calledAfterViewInjection() {
//Other variables initaialization

}

 @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
       
        final AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());
        final LayoutInflater inflater = getActivity().getLayoutInflater();
        view = inflater.inflate(R.layout.activity_location_view, null);
        builder.setView(view).setPositiveButton("OK", new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {

                    }
                })
                .setNegativeButton("CANCEL", (dialog, id) -> LocationViewActivity.this.getDialog().cancel());

        return builder.create();

    }
`

**NOTE ** 

- `@EFragment(R.layout.activity_location_view)` and ` view = inflater.inflate(R.layout.activity_location_view, null);` are needed for dialog to load @EFragment content.
- To Load DialogFragment call. `DialogFragment dialog = new LocationViewActivity_();
        dialog.show(getSupportFragmentManager(), "dialog");`
- In the above, note DialogFragment object. If you make LocationViewActivity_ object you get Fragment not attached to Activity exception 

**androidannotations is a great framework though learning is steep**
